### PR TITLE
Vue3:implement inheritAttrs config for vue3 renderer

### PR DESCRIPTION
**Pull request by chakAs3**
Closes #

this PR closes issues for v6 stories that not using fallthough attributes 

## What I did

I have added inheritAttrs as parameter can be set to override the default behavor "fallthrough attributes" starting from vue3.0.0.

users can globally disable this behavior if it is not disared for all stories ( in preview ), or component level in meta.
or in Story level.
user can disable this also without using context.parameters by  just set inheritAttrs = false in the component returned by the render function 

Order by priority 

### in Story render component 
```ts
export const CustomRender: Story = {
  render: (args: any) => ({
    components: { MySlotComponent },
    inheritAttrs: true, // by setting this to true, we automatically passing the args to the component as attributes
    // we don't need to pass label and year as props to the component because we enabled inheritAttrs in this component
    template: `<Component   onEvent='eventHandler'>
  	             some custom text
              </Component>`,
  }),
};
```

### in Story parameters 
```ts
export const CustomRender: Story = {
  render: (args: any) => ({
    components: { MySlotComponent },
    // we don't need to pass label and year as props to the component because we enabled inheritAttrs in this component
    template: `<Component   onEvent='eventHandler'>
  	             some custom text
              </Component>`,
  }),
 parameters : { inheritAttrs: true } // by setting this to true, we automatically passing the args to the component as attributes
};
```

### in Preview parameters ( all the stories)
```ts
// .storybook/preview.js
 const preview = {
  parameters: {
    inheritAttrs: false, // disable global inheritAttrs for all stories
  },
};

```

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

**Comment by chakAs3** [comment]
@tmeasday  i need your feedback since you are behind the rendering mechanism.
my idea is basically:  Benefit of story render  
```ts
( args:StoryArgs ) => VueRenderer 
```

and provide to Vue user the same api and behavior should a VueRenderer have ( Functional Component, Concret component options or composition ). it is the case now i think for all renderers.  However in v6 that was some PR that just disabled some Vue3.0.0 main feature cause it is not desirable for the specific user. 

the right way from point of view is the preserve the Vue api and behavior, and provide global config using context.parameters to override this behavior whether globally in preview.js , or in the Story level

File: code/renderers/vue3/template/stories/InheritAttrs.stories.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  component: InheritAttrs,
+  argTypes: {},
+  render: (args: any) => ({
+    // Components used in your story `template` are defined in the `components` object
+    components: { InheritAttrs },
+    // The story's `args` need to be mapped into the template through the `setup()` method
+    setup() {
+      return { args };
+    },
+    // And then the `args` are bound to your component with `v-bind="args"`
```
Comment (bodograumann) [inline code/renderers/vue3/template/stories/InheritAttrs.stories.ts:17]: There is no `v-bind="args"` here though.

Comment (bodograumann) [reply code/renderers/vue3/template/stories/InheritAttrs.stories.ts:17]: Also using `<InheritAttrs />` should be preferred here imho.

File: code/renderers/vue3/template/stories/InheritAttrs.stories.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    template: '<inherit-attrs  />',
```
Comment (bodograumann) [inline code/renderers/vue3/template/stories/InheritAttrs.stories.ts:18]: Using <InheritAttrs /> should be preferred here imho.

**Comment by bodograumann** [comment]
(Picking up the discussion from https://github.com/storybookjs/storybook/issues/21379)

First of all, it is not quite clear what this feature is supposed to achieve. You say that "it is not template that users set in the render function it is a component". That is not completely correct though. The render function is a component factory function which is given the args and the storybook context. So what reason could there be to expect that the args are actually *also* applied as attributes to the component returned from the render function?

Secondly I would like to understand what practical advantage passing args this way has. Here is my thinking:

When would this feature work?
- If we render the documented component as a single root node.
  ```
  (args) => ({
    components: { Button },
    template: `<Button>label</Button>`,
  }),
  ```

It would however not work
- when we render multiple root nodes
  ```
  (args) => ({
    components: { Button },
    template: `
      <Button>label</Button>
      <hr />
    `,
  }),
  ```
- when we wrap the component in a plain html element (directly or through some decorator)
  ```
  (args) => ({
    components: { Button },
    template: `<div class="bg-black text-white p-3">
      <Button>label</Button>
    </div>`,
  }),
  ```

I guess especially the last variant is quite common. For example it is necessary when the documented component (e.g. Button) renders multiple root nodes and you want to add some padding in the storybook.

So in the first situation it might make writing the render function slightly more succinct, but in the other cases it would mean having to add `inheritAttrs: false`. Forgetting about that might lead to hard-to-understand errors like in https://github.com/storybookjs/storybook/issues/21379 .
On the other hand without the new feature a missing `v-bind="args"` is obvious because the args will not be applied at all.

In the end, the user will have to learn how to properly use args in his render function anyway.
So from my point of view this feature does not make much sense.

**Now on to more technical things**:

You are quite far behind `next`. I think you will find that after rebasing args will not be applied as attributes anymore. So you should probably add that back. (Cf. https://github.com/storybookjs/storybook/pull/22229/files)
What confused me for bit was that `inheritAttrs` defaults to `true` in vue. (Seems like bad practice to me, but I guess we have to live with it now. Usually you would use something like `noInheritAttrs` instead.)
Here in the storybook, that means every existing user has it enabled by default. Adding `inheritAttrs: false` to the template does not change that for those exsting users. Using a migration for the existing users for an experimental feature also seems like a bad idea. So we need a feature flag that defaults to `false`. Your parameter is not such a thing, because it defaults to `undefined` and that means `true` as an option on the story component.

Looking at the PR description your examples use `MySlotComponent` and `Component`. They are supposed to be the same right? Not sure what meaning the "slot" has here. Maybe replace both with `Button`. Also `Component` is a dynamic component in vue: https://vuejs.org/api/built-in-special-elements.html#component


**As a final note, here are some things that I think would actually help**:

When I looked back at the documentation to get a better picture what way of writing stories is taught "these days" :-D, I found it quite overwhelming.
I think introducing the render function as early as done in the docs currently is a mistake.
There is actually a default render function that should work fine for the beginner and which does apply all args as vue props without the need for `inheritAttrs` (Cf. https://github.com/storybookjs/storybook/blob/next/code/renderers/vue3/src/render.ts#L9-L18).
So this definition of stories, just with args and without a render function, should be the first thing the user learns through the documentation.
Later more advanced techniques can be taught, including a custom render function. Most-importantly it should be shown how the render function can be reused in all stories for a component.
And there we can easily show how to pass args to the component with `setup: () => ({ args })` and `v-bind="args"` in the template.

**Comment by shilman** [comment]
Thanks so much for the comments @bodograumann! Indeed, we're going to close this PR and pursue @chakAs3 's proposal for how to handle non-React custom stories as part of a larger, long-term project for Storybook. I've [started a rough draft](https://www.notion.so/chromatic-ui/RFC-Vue-story-components-9dc085671df34e12901b48fcba4b491d?pvs=4) for discussion and will release it as a formal RFC once @chakAs3 @kasperpeulen @tmeasday and the other maintainers have weighed in.

**Comment by kasperpeulen** [comment]
@bodograumann I actually completely agree with your points including the docs, I started a PR for that here:
https://github.com/storybookjs/storybook/pull/22831

**Comment by chakAs3** [comment]
> "it is not template that users set in the render function it is a component"

hi @bodograumann, I did not say that the render function is a component,  it is already a function so it will not satisfy all Vue component type,  what I said : > "it is function that returns a Component". the idea here is to understand that the return from your render function should be a Vue Component that will render your story.

**Pull request closed without merging**